ddtrace/contrib/datadog_lambda/patch.py

```diff
@@ -0,0 +1,120 @@
+    """Raises an `ImpendingTimeout` error which causes the tracer to shutdown."""
+    # TODO: Update to tag error in local root parent instead of raising the error.
+
+    raise ImpendingTimeout("Datadog detected an Impending Timeout")
+
+
+def check_timeout(context):
+    """
+    Creates a timeout to detect when an AWS Lambda handler's remaining
+    time is about to end.
+
+    Crashes flushes when the signal is activated
+    """
+    signal.signal(signal.SIGALRM, crash_flush)
+    remaining_time_in_millis = context.get_remaining_time_in_millis()
+    apm_flush_deadline = int(os.environ.get("DD_APM_FLUSH_DEADLINE", 0))
+    if apm_flush_deadline and apm_flush_deadline <= remaining_time_in_millis:
+        remaining_time_in_millis = apm_flush_deadline
+
+    remaining_time_in_seconds = (remaining_time_in_millis - 100) / 1000
+    signal.setitimer(signal.ITIMER_REAL, remaining_time_in_seconds, 0)
+
+
+def datadog_instrumentation(func, args, kwargs):
+    """Patches an AWS Lambda handler function for Datadog instrumentation."""
+    context = args[-1]  # context is always the last parameter
+    check_timeout(context)
+
+    return func(*args, **kwargs)
+
+
+def modify_module_name(module_name):
+    """Returns a valid modified module to get imported."""
+    return ".".join(module_name.split("/"))
+
+
+def get_handler_and_module():
+    """Returns the user AWS Lambda handler and module."""
+    path = os.environ.get("DD_LAMBDA_HANDLER", None)
+    if path is None:
+        from datadog_lambda.wrapper import datadog_lambda_wrapper
+
+        handler = getattr(datadog_lambda_wrapper, "__call__")
+
+        def wrapper(func, args, kwargs):
+            return datadog_instrumentation(func, args, kwargs)
+
+        return handler, datadog_lambda_wrapper, wrapper
+    else:
+        parts = path.rsplit(".", 1)
+        (mod_name, handler_name) = parts
+        modified_mod_name = modify_module_name(mod_name)
+        handler_module = import_module(modified_mod_name)
+        handler = getattr(handler_module, handler_name)
+        return handler, handler_module, datadog_instrumentation
+
+
+def _has_patch_module():
+    """
+    Ensures that the `datadog_lambda` integration can be patched.
+
+    It checks either the user has the DD_LAMBDA_HANDLER set correctly.
+    Or if the `datadog_lambda` package is installed.
+    """
+    path = os.environ.get("DD_LAMBDA_HANDLER", None)
+    if path is None:
+        try:
+            import_module("datadog_lambda.wrapper")
+        except Exception:
+            return False
+    else:
+        parts = path.rsplit(".", 1)
+        if len(parts) != 2:
+            return False
+    return True
+
+
+def _in_aws_lambda():
+    """Checks if environment is an AWS Lambda handler."""
+    return bool(os.environ.get("AWS_LAMBDA_FUNCTION_NAME", False))
+
+
+def patch():
+    """Patches an AWS Lambda using the `datadog-lambda-py` Lambda layer."""
+    if not _in_aws_lambda() and not _has_patch_module():
```

Comment:
Under what circumstances outside of AWS lambda do you intend this patch to happen?

Reply:
I don't intend for this patch to happen in any other situation.

Reply:
In that case I don't understand why there's an `and` here - could you get by with only the `not _in_aws_lambda` check?

Reply:
Yes, I could just get by only using it. Although I'm avoiding here for the user to have only the `ddtrace` package installed without the `datadog-lambda` package. Let's say a user has a Lambda function with `ddtrace`, instead of having the tracer with the Datadog Lambda Extension, its tracer is configured to a remote agent or pointed to an EC2 instance running it through a VPC. The code will fail in this case, since `get_handler_and_module` assumes the user instrumented its Lambda function with Datadog.

Technically possible, although a really obscure case. 

What do you think about this?

Reply:
@duncanista to be honest I don't think I have enough context to follow what you're saying. This is a pretty minor point that's mostly about code readability, so I think if you make it clear in a comment why this logic exists, it's fine to keep it.